akka-actor/src/main/scala/akka/dispatch/BatchingExecutor.scala

```diff
@@ -78,22 +78,23 @@ private[akka] trait BatchingExecutor extends Executor {
     }
   }
 
+  private[this] val _blockContext = new ThreadLocal[BlockContext]()
+
   private[this] final class BlockableBatch extends AbstractBatch with BlockContext {
-    private var parentBlockContext: BlockContext = _
     // this method runs in the delegate ExecutionContext's thread
     override final def run(): Unit = {
       require(_tasksLocal.get eq null)
       _tasksLocal set this // Install ourselves as the current batch
-      val prevBlockContext = BlockContext.current
+      val prevBlockContext = _blockContext.get
+      _blockContext.set(BlockContext.current)
       BlockContext.withBlockContext(this) {
```

Comment:
Doesn't `withBlockContext` already handle `prev` itself?

Also, when you fix this bug you'll want to forward-port it to Scala: https://github.com/scala/scala/blob/2.11.x/src/library/scala/concurrent/BatchingExecutor.scala